### ISSUE TITLE
Mutation, Composition and Utility Value Generators

### ISSUE THREAD

**Issue by Aatch** [brendanzab/noise-rs#100]
I propose adding a new, higher-level API for constructing complex noise functions built from simpler parts. There are three parts to this: mutators that alter either the input value or the output value, composition functions that allow multiple value sources to be combined in interesting ways and "utility" value generators.

## Mutators

Mutators are probably the simplest part. They wrap a noise source and can choose to alter either the input point, the generated value or both. This means mutators can be used for a variety of tasks, from rotating a value source to scaling output values.

## Composition

Composition functions can be used combine value sources. The general pattern here is that they are constructed with multiple value sources and can combine them in interesting ways. For example, blending between two value sources dependent on a third. This could be used for biome generation where one layer determines the biome and different biomes have different terrain generation functions.

## Utility Generators

These are special value sources that produce some specific result. The simplest one is a constant-value generator that always returns the same value no matter what. More interesting is checkerboard, striped or similar patterns. The main reason for having these is to supply values to be manipulated using noise in some way.

## General Design

The general design I envision is something similar to `Iterator`. Having a trait for each dimension that just takes a point and returns a value. This allows for users to do whatever they want. The library would then provide some implementers of the trait that do some common tasks like mapping inputs, mapping outputs and the utility generators mentioned.

This is not intended to replace the existing API, just build upon it.

**Comment by bjz** [comment]
Sounds cool! It would be helpful to have some type signatures to get a clearer sense of the API you are proposing.

**Comment by Aatch** [comment]
The core traits would be:

```rust
trait ValueSourceN<T> {
    type Output;

    fn get(&mut self, pnt: &PointN<T>) -> Self::Output;
}
```

for N = 1,2,3,4. The placement of the `T` parameter for the point types is a tricky decision. While it makes more sense for it to go on the `get` method, that prevents the trait from being object-safe. It could be an associated type, but that doesn't seem like a good fit, semantically. Another point is whether or not `self` should be mutable or not. Generally, it shouldn't need to be, but it is the most flexible choice in terms of trait implementation. However it does restrict usage (can't use the same `ValueSource` for multiple pieces in a composite `ValueSource`).

Names are obviously up for debate, I don't really care much. A sketch implementation for a composer is this:

```rust
struct Select2<T, I1: ValueSource2<T>, I2: ValueSource2<T>, S: ValueSource2<T>> {
    input_1: I1,
    input_2: I2,
    selector: S
}

impl<I1, I2, S, T> ValueSource2<T> for Select2<T, I1, I2, S, T>
where I1: ValueSource2<T>,
           I2: ValueSource2<T, Output=I1::Output>,
           S: ValueSource2<T>, S::Output : Float
{
    type Output = I1::Output;

    fn get(&mut self, pnt: &Point2<T>) -> Self::Output {
        let select = self.selector.get(pnt);

        let half = math::cast(0.5);
        if select < half {
            self.input_1.get(pnt)
        } else {
            self.input_2.get(pnt)
        }
    }
}
```

Assuming I didn't screw up too much stuff, it should get the idea across, this just uses a noise source to select between two alternate input sources. A real implementation would probably allow the definition of some sort of blending function instead of a simplistic, hard-coded cut-off.

**Comment by Aatch** [comment]
Hmm, I thought up another possible form for the base trait:

```rust
pub trait ValueSource {
    type Input;
    type Output;

    fn get(&self, pnt: &Self::Input) -> Self::Output;
}
```

This makes the input type an associated type, meaning that dimension-agnostic functionality like the `Select` type above can just pass along the input point blindly. Using it generically isn't too bad:

```rust
fn do_stuff<T: Float, V: ValueSource<Input=Point2<T>>>(vals: &V)
where V::Output : Float
{ ... }
```

And that's more generic than I'd probably expect for most people, user code is more likely to hardcode the types to `f32` or `f64` instead.

Allowing for basically any input seems a little *too* flexible, but I can't think of a good reason why it shouldn't be allowed. I've done some basic tests to make sure the code works and doesn't hit any major type-system issues and it works perfectly so far.

**Comment by Aatch** [comment]
I started to implement it and have run into a design question: should the `Seed` be part of the state of of the source or passed to the `get` method. If it's part of the `ValueSource`, should it be `&Seed` or `Seed`?

**Comment by Aatch** [comment]
Another question is how to deal with `BrownianN`, it kinda falls more into this category in that it modifies a raw noise source. Depending on how the previous question is answered, this might not matter too much, since `Brownian` could just impl `ValueSource` directly without changes, would require a seed to be stored, or require a wrapper that stores the seed.

**Comment by Aatch** [comment]
I've started on a proof-of-concept implementation and made a `Select` type that selects between two noise sources with the option of blending the edges. Which produces output like this:

![source](https://cloud.githubusercontent.com/assets/342416/12969739/e83d171c-d0e9-11e5-8d8f-03154b88f0b0.png)

**Comment by bjz** [comment]
Sorry for the late response.

With `Brownian`, I'm wondering if `F` could be converted to a `ValueSource`?

**Comment by bjz** [comment]
Feel free to make a WIP PR by the way, would be great to see the progress.

**Comment by Aatch** [comment]
@bjz ah yeah, I was kinda waiting on some answers to my questions before, so I stopped working on it and then it just fell off of my radar. The main one is handling the `Seed` parameter. My current implementation uses a stored `&Seed` in the `Source` implementations for the base sources (perlin, open simplex).

**Comment by bjz** [comment]
I'm guessing `&Seed`. Feel free to run with this and com up with a design - I don't want to hold you up!

**Comment by Razaekel** [comment]
@Aatch are you still working on this? I'd like to contribute.

**Comment by bjz** [comment]
@Aatch: any updates?

**Comment by Razaekel** [comment]
After working my way though it, I've produced this example setup, available on the rust playground:

https://is.gd/JUbUEi

It basically ended up the same as Aatch's first ValueSource trait, minus the N on the end.

The goal was to achieve an API where the user could construct a chain of modules, and then simply call get(PointN) on the last module, irrespective of what module it actually is, and the whole thing would pass the PointN through their own get() functions to resolve the chain.

This should allow the user to simply change their PointN type and the whole chain will 'just work' for the new type, as long as the PointN type is one that's been impl'd on all the modules in the chain.

**Comment by bjz** [comment]
Looks good to me!

**Comment by bjz** [comment]
I wonder how it would look with the other noise generators...

**Comment by Razaekel** [comment]
Closing this, since it's being implemented as part of the milestone for v0.3.0.
